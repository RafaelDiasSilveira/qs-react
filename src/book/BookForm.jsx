import React from 'react';
import PropTypes from 'prop-types';

import moment from 'moment';

import 'react-dates/initialize';
// import { SingleDatePicker } from 'react-dates';

class BookForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      titulo: props.book ? props.book.titulo : '',
      autor: props.book ? props.book.autor : '',
      dataPublicacao: props.book ? props.book.dataPublicacao : '',
      genero: props.book ? props.book.genero : '',
      error: '',
    };
  }

  onTituloChange = (e) => {
    const titulo = e.target.value;
    this.setState(() => ({ titulo }));
  };

  onAutorChange = (e) => {
    const autor = e.target.value;
    this.setState(() => ({ autor }));
  };

  onDateChange = (e) => {
    const dataPublicacao = e.target.value;
    this.setState(() => ({ dataPublicacao }));
  }

  onGeneroChange = (e) => {
    const genero = e.target.value;
    this.setState(() => ({ genero }));
  }

  onSubmit = (e) => {
    e.preventDefault();

    let error;

    if (!this.state.titulo) {
      error = 'Título inválido';
      this.setState(() => ({ error }));
    } else {
      this.props.onSubmit({
        titulo: this.state.titulo,
        autor: this.state.autor,
        dataPublicacao: this.state.dataPublicacao,
        genero: this.state.genero,
      });
    }
  };

  render() {
    return (
      <div>

        { this.state.error &&
          (
            <div className="alert alert-danger" role="alert">
              Erros encontrado:
              {this.state.error}
            </div>
          )
        }

        <form onSubmit={this.onSubmit}>

          <div className="form-row">
            <div className="col form-group">
              <label>Título</label>
              <input
                type="text"
                placeholder="Titulo"
                value={this.state.titulo}
                onChange={this.onTituloChange}
                className="form-control"
              />
            </div>

            <div className="col form-group">
              <label>Autor</label>
              <input
                type="text"
                placeholder="Autor"
                value={this.state.autor}
                onChange={this.onAutorChange}
                className="form-control"
              />
            </div>
          </div>

          <div className="form-row">
            <div className="col form-group">
              <label htmlFor="generoSelect">
              Gênero
              </label>
              <select
                id="generoSelect"
                className="form-control"
                value={this.state.genero}
                onChange={this.onGeneroChange}
              >
                <option value="" disabled selected> -- selecione uma opção -- </option>
                <option value="epopeia">EPOPEIA</option>
                <option value="novela">NOVELA</option>
                <option value="conto">CONTO</option>
                <option value="ensaio">ENSAIO</option>
                <option value="romance">ROMANCE</option>
              </select>
            </div>

            <div className="col form-group">
              <label>Data de Publicação</label>
              <input
                type="date"
                className="form-control"
                value={this.state.dataPublicacao}
                onChange={this.onDateChange}
                defaultValue="31/12/18"
              />
            </div>
          </div>

          <button className="btn btn-success btn-lg btn-block">Adicionar Livro</button>

        </form>
      </div>
    );
  }
}

BookForm.propTypes = {
  book: PropTypes.shape({
    titulo: PropTypes.string,
    autor: PropTypes.string,
    dataPublicacao: PropTypes.instanceOf(moment),
    genero: PropTypes.string,
  }),
  onSubmit: PropTypes.func.isRequired,
};

BookForm.defaultProps = {
  book: undefined,
};

export default BookForm;
