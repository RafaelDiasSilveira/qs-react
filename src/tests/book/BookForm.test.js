import React from 'react';
import { shallow } from "enzyme";
import moment from 'moment';

import BookForm from '../../book/BookForm';
import { books } from "../fixtures/fixtures";

let wrapper, onSubmit;

beforeAll(() => {
  onSubmit = jest.fn;
  wrapper = shallow(<BookForm onSubmit={onSubmit} />);

});

describe('should render BookForm correctly', () => {
  test('with data', () => {
    const wrapper = shallow(
      <BookForm
        book={books[0]}
        onSubmit={onSubmit}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  test('without data', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('should set state correctly on input change', () => {
  test('when setting titulo input', () => {
    const value = 'Teste do GG';

    wrapper.find('input').at(0).simulate('change', {
      target: { value }
    });

    expect(wrapper.state('titulo')).toBe(value);
  });

  test('when setting autor input', () => {
    const value = 'Teste do GG';

    wrapper.find('input').at(1).simulate('change', {
      target: { value }
    });

    expect(wrapper.state('autor')).toBe(value);
  });

  test('when setting dataPublicacao input', () => {
    const now = moment("2018-12-25");

    wrapper.find('input').at(2).simulate('change', {
      target: { value: now }
    });

    expect(wrapper.state('dataPublicacao')).toBe(now);
  });

  test('when setting genero input', () => {
    const genero = 'romance';

    wrapper.find('select').at(0).simulate('change', {
      target: { value: genero }
    });

    expect(wrapper.state('genero')).toBe(genero);
  });
})


