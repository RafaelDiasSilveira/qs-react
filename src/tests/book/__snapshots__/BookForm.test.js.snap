// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BookForm correctly with data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-row"
    >
      <div
        className="col form-group"
      >
        <label>
          Título
        </label>
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="Titulo"
          type="text"
          value="Livro 1"
        />
      </div>
      <div
        className="col form-group"
      >
        <label>
          Autor
        </label>
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="Autor"
          type="text"
          value="Autor 1"
        />
      </div>
    </div>
    <div
      className="form-row"
    >
      <div
        className="col form-group"
      >
        <label
          htmlFor="generoSelect"
        >
          Gênero
        </label>
        <select
          className="form-control"
          id="generoSelect"
          onChange={[Function]}
          value="EPOPEIA"
        >
          <option
            disabled={true}
            selected={true}
            value=""
          >
             -- selecione uma opção -- 
          </option>
          <option
            value="epopeia"
          >
            EPOPEIA
          </option>
          <option
            value="novela"
          >
            NOVELA
          </option>
          <option
            value="conto"
          >
            CONTO
          </option>
          <option
            value="ensaio"
          >
            ENSAIO
          </option>
          <option
            value="romance"
          >
            ROMANCE
          </option>
        </select>
      </div>
      <div
        className="col form-group"
      >
        <label>
          Data de Publicação
        </label>
        <input
          className="form-control"
          defaultValue="31/12/18"
          onChange={[Function]}
          type="date"
        />
      </div>
    </div>
    <button
      className="btn btn-success btn-lg btn-block"
    >
      Adicionar Livro
    </button>
  </form>
</div>
`;

exports[`should render BookForm correctly without data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-row"
    >
      <div
        className="col form-group"
      >
        <label>
          Título
        </label>
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="Titulo"
          type="text"
          value=""
        />
      </div>
      <div
        className="col form-group"
      >
        <label>
          Autor
        </label>
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="Autor"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="form-row"
    >
      <div
        className="col form-group"
      >
        <label
          htmlFor="generoSelect"
        >
          Gênero
        </label>
        <select
          className="form-control"
          id="generoSelect"
          onChange={[Function]}
          value=""
        >
          <option
            disabled={true}
            selected={true}
            value=""
          >
             -- selecione uma opção -- 
          </option>
          <option
            value="epopeia"
          >
            EPOPEIA
          </option>
          <option
            value="novela"
          >
            NOVELA
          </option>
          <option
            value="conto"
          >
            CONTO
          </option>
          <option
            value="ensaio"
          >
            ENSAIO
          </option>
          <option
            value="romance"
          >
            ROMANCE
          </option>
        </select>
      </div>
      <div
        className="col form-group"
      >
        <label>
          Data de Publicação
        </label>
        <input
          className="form-control"
          defaultValue="31/12/18"
          onChange={[Function]}
          type="date"
          value=""
        />
      </div>
    </div>
    <button
      className="btn btn-success btn-lg btn-block"
    >
      Adicionar Livro
    </button>
  </form>
</div>
`;
